torch/testing/_internal/common_utils.py

```diff
@@ -777,27 +777,65 @@ def safeCoalesce(self, t):
 
         return tg
 
-    def assertEqual(self, x, y, prec=None, message='', allow_inf=False, exact_dtype=None):
+    # Some analysis of tolerance by logging tests from test_torch.py can be found
+    # in https://github.com/pytorch/pytorch/pull/32538.
+    dtype_precisions = {
+        torch.half: 0.001,
+        torch.bfloat16: 0.016,
+        torch.float: 1.3e-6,
+        torch.double: 5e-7,
+        torch.int: 0
+    }
+
+    def compared_dtype(self, x, y):
+        dtype = torch.float
+        if isinstance(x, torch.Tensor):
+            dtype = x.dtype
+        elif isinstance(y, torch.Tensor):
+            dtype = y.dtype
+        elif TEST_NUMPY and isinstance(x, numpy.ndarray):
+            dtype = torch.from_numpy(x).dtype
+
+        if dtype.is_floating_point:
+            return dtype
+        else:
+            return torch.int
+
+
+    def assertEqual(self, x, y, message='', rtol=None, atol=None, allow_inf=False, exact_dtype=None):
+        # we allow setting an absolute tolerance as a positional arg for BC with legacy testing behavior.
```

Comment:
It might be easier if you do a preparatory commit before this one that makes everything after `tol` kwarg only, and then fix call sites.

Reply:
This is going to need XLA updates too.

Reply:
actually I did, I forgot that wasn't allowed in python2. I had it working that way, and then changed it back this morning as py2 tests failed.

Reply:
and, yeah I noticed that XLA tests failed. cc @ailzhang

Reply:
If you go kwarg only you're going to have to do it manually; take `**kwargs` and then parse it out. Well, or wait until we kill py2 support.